llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -6903,18 +7031,97 @@ SDValue RISCVTargetLowering::LowerOperation(SDValue Op,
+      int Log2LMUL = Log2_64(NumElts) - 3;
+
+      auto Flag = SDNodeFlags();
+      Flag.setNoUnsignedWrap(true);
+      SDValue Ret = DAG.getUNDEF(VecTy);
+      SDValue BasePtr = Load->getBasePtr();
+      if (auto *FI = dyn_cast<FrameIndexSDNode>(BasePtr))
+        MFI.setStackID(FI->getIndex(), TargetStackID::ScalableVector);
+      SDValue VROffset = DAG.getNode(RISCVISD::READ_VLENB, DL, XLenVT);
+      VROffset =
+          DAG.getNode(ISD::SHL, DL, XLenVT, VROffset,
+                      DAG.getConstant(std::max(Log2LMUL, 0), DL, XLenVT));
+
+      // Load NF vector registers and combine them to a vector tuple.
+      for (unsigned i = 0; i < NF; ++i) {
+        SDValue LoadVal = DAG.getLoad(
+            MVT::getScalableVectorVT(MVT::i8, NumElts), DL, DAG.getEntryNode(),
+            BasePtr, MachinePointerInfo(Load->getAddressSpace()), Align(8));
+        Ret = DAG.getNode(RISCVISD::INSERT_SUBVECTOR, DL, VecTy, Ret, LoadVal,
+                          DAG.getVectorIdxConstant(i, DL));
+        BasePtr = DAG.getNode(ISD::ADD, DL, XLenVT, BasePtr, VROffset, Flag);
+      }
+      return DAG.getMergeValues({Ret, DAG.getEntryNode()}, DL);
+    }
+
     if (auto V = expandUnalignedRVVLoad(Op, DAG))
       return V;
     if (Op.getValueType().isFixedLengthVector())
       return lowerFixedLengthVectorLoadToRVV(Op, DAG);
     return Op;
-  case ISD::STORE:
+  }
+  case ISD::STORE: {
+    auto *Store = cast<StoreSDNode>(Op);
+    SDValue StoredVal = Store->getValue();
+    EVT VecTy = StoredVal.getValueType();
+    // Handle normal vector tuple store.
+    if (VecTy.isRISCVVectorTuple()) {
+      MachineFunction &MF = DAG.getMachineFunction();
+      MachineFrameInfo &MFI = MF.getFrameInfo();
+      SDLoc DL(Op);
+      MVT XLenVT = Subtarget.getXLenVT();
+      unsigned NF = VecTy.getRISCVVectorTupleNumFields();
+      unsigned Sz = VecTy.getSizeInBits();
+      unsigned NumElts = Sz / (NF * 8);
+      int Log2LMUL = Log2_64(NumElts) - 3;
+
+      auto Flag = SDNodeFlags();
+      Flag.setNoUnsignedWrap(true);
+      SDValue Ret;
+      SDValue Chain = Store->getChain();
+      SDValue BasePtr = Store->getBasePtr();
+      if (auto *FI = dyn_cast<FrameIndexSDNode>(BasePtr))
+        MFI.setStackID(FI->getIndex(), TargetStackID::ScalableVector);
+      SDValue VROffset = DAG.getNode(RISCVISD::READ_VLENB, DL, XLenVT);
+      VROffset =
+          DAG.getNode(ISD::SHL, DL, XLenVT, VROffset,
+                      DAG.getConstant(std::max(Log2LMUL, 0), DL, XLenVT));
+
+      // Extract subregisters in a vector tuple and store them individually.
+      for (unsigned i = 0; i < NF; ++i) {
+        auto Extract = DAG.getNode(RISCVISD::EXTRACT_SUBVECTOR, DL,
+                                   MVT::getScalableVectorVT(MVT::i8, NumElts),
+                                   StoredVal, DAG.getVectorIdxConstant(i, DL));
+        Ret = DAG.getStore(Chain, DL, Extract, BasePtr,
+                           MachinePointerInfo(Store->getAddressSpace()),
```

Comment:
Why is this discarding the pointer info?

Reply:
The correct pointer info needs a scalable offset, but that doesn't seem  to be supported yet. Is there something better we can do?

Reply:
I thought that got added?

Reply:
The offset field in MachinePointerInfo is still an `int64_t` and `getWithOffset` takes a `int64_t`. SelectionDAGBuilder::visitLoad/visitStore still have this code.

```
    // TODO: MachinePointerInfo only supports a fixed length offset. 
    MachinePointerInfo PtrInfo =                                                 
        !Offsets[i].isScalable() || Offsets[i].isZero()                          
            ? MachinePointerInfo(PtrV, Offsets[i].getKnownMinValue())            
            : MachinePointerInfo();
```